### ISSUE TITLE
Robustness issues of CLI commands and services

### ISSUE THREAD

**Issue by nearmax** [near/rainbow-bridge#142]
Many of our CLI commands have transient failures because we don't take care of failures that can happen when interacting with Ethereum or NEAR node. The most common sources of failures are:
1. Broken connection that was not re-established. Usually happens in near-api-js, and can manifest as either ECONRESET or another type of error message, see below;
2. Concurrent access key usage on NEAR side or concurrent key usage on Ethereum side. Since every time the key is used it needs to be used with a new nonce, what happens if two services try using the same key with the same nonce simultaneously one transaction will fail. Ideally, we should be able to recognize this failure and retry again;
3. Incorrect nonce tracking. Sometimes even when the key is not used concurrently, both web3js and near-api-js would attempt to use an incorrect nonce, for unknown reason. We should make sure we have a workaround for this issue.
4. Some contract initializations are not robust. E.g. init-near-contracts will fail if the account was already created but the code was not deployed yet.

We need to fix the above scenarios for each of the CLI commands, starting with the commands that start daemons, because daemons should be extremely fault-proof for the bridge to work smoothly.

Additionally, we should consider using pm2's feature of restarting the service when it fails.

Attaching examples of CLI command failures below:
<details>
<summary>near-api-js for whatever reason uses wrong nonce. Case (3). </summary>
node index.js transfer-eth-erc20-to-near --amount 1000 --eth-sender-sk 0x2bdd21761a483f71054e14f5b827213567971c676928d9a1808cbfa4b7501200 --near-receiver-account eth2nearprover --near-master-account nearfuntoken

Approving token transfer.
Approved token transfer.
Transferring tokens from the ERC20 account to the token locker account.
Success.
Transferring 1000 tokens from 0x7cC4B1851c35959D34e635A470F6b5C43bA3C9c9 ERC20. From 0xDf08F82De32B8d460adbE8D72043E3a7e25A3B39 sender to eth2nearprover recipient
Eth2NearClient is currently at block 7. Waiting for block 12 to be confirmed. Sleeping for 10 sec.
Eth2NearClient is currently at block 8. Waiting for block 12 to be confirmed. Sleeping for 10 sec.
Eth2NearClient is currently at block 10. Waiting for block 12 to be confirmed. Sleeping for 10 sec.
Eth2NearClient is currently at block 11. Waiting for block 12 to be confirmed. Sleeping for 10 sec.
Eth2NearClient is currently at block 12. Waiting for block 12 to be confirmed. Sleeping for 10 sec.
Eth2NearClient is currently at block 14. Waiting for block 12 to be confirmed. Sleeping for 10 sec.
Eth2NearClient is currently at block 15. Waiting for block 12 to be confirmed. Sleeping for 10 sec.
Eth2NearClient is currently at block 17. Waiting for block 12 to be confirmed. Sleeping for 10 sec.
Eth2NearClient is currently at block 18. Waiting for block 12 to be confirmed. Sleeping for 10 sec.
Eth2NearClient is currently at block 19. Waiting for block 12 to be confirmed. Sleeping for 10 sec.
Eth2NearClient is currently at block 20. Waiting for block 12 to be confirmed. Sleeping for 10 sec.
Eth2NearClient is currently at block 21. Waiting for block 12 to be confirmed. Sleeping for 10 sec.
Eth2NearClient is currently at block 21. Waiting for block 12 to be confirmed. Sleeping for 10 sec.
Balance of eth2nearprover before the transfer is 0
TxHash HFpfDDiuCWR5qKS3jQgm9KTjNGCgvkJ9bXyXzHqoUV1C
Failed:  { Error: [-32000] Server error: Invalid transaction: Transaction nonce 3 must be larger than nonce of the used access key 3
    at JsonRpcProvider.sendJsonRpc (/Users/maksymzavershynskyi/Projects/rainbow-bridge/environment/node_modules/near-api-js/lib/providers/json-rpc-provider.js:145:27)
    at process._tickCallback (internal/process/next_tick.js:68:7) type: 'UntypedError' }
(node:15534) UnhandledPromiseRejectionWarning: Error: [-32000] Server error: Invalid transaction: Transaction nonce 3 must be larger than nonce of the used access key 3
    at JsonRpcProvider.sendJsonRpc (/Users/maksymzavershynskyi/Projects/rainbow-bridge/environment/node_modules/near-api-js/lib/providers/json-rpc-provider.js:145:27)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:15534) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:15534) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
</details>
<details>
<summary>eth-relay fails with ECONRESET Case (1). </summary>
1|eth-relay  | Successfully submitted block 14 to EthClient
1|eth-relay  | { FetchError: request to http://localhost:3030/ failed, reason: read ECONNRESET
1|eth-relay  |     at ClientRequest.<anonymous> (/Users/maksymzavershynskyi/Projects/rainbow-bridge/environment/node_modules/node-fetch/lib/index.js:1455:11)
1|eth-relay  |     at ClientRequest.emit (events.js:198:13)
1|eth-relay  |     at Socket.socketErrorListener (_http_client.js:401:9)
1|eth-relay  |     at Socket.emit (events.js:198:13)
1|eth-relay  |     at emitErrorNT (internal/streams/destroy.js:91:8)
1|eth-relay  |     at emitErrorAndCloseNT (internal/streams/destroy.js:59:3)
1|eth-relay  |     at process._tickCallback (internal/process/next_tick.js:63:19)
1|eth-relay  |   message:
1|eth-relay  |    'request to http://localhost:3030/ failed, reason: read ECONNRESET',
1|eth-relay  |   type: 'system',
1|eth-relay  |   errno: 'ECONNRESET',
1|eth-relay  |   code: 'ECONNRESET' }
</details>
<details>
<summary>web3js for whatever reason fails to open connection. Case (1). </summary>
node index.js init-near-contracts
Key to call Eth2NearClient contract is not specified. Reusing master key.
Key to call Eth2NearProver contract is not specified. Reusing master key.
Creating accounts and deploying the contracts.
Account eth2nearclient does not exist creating it.
Created account eth2nearclient
Deployed contract to account eth2nearclient
Account eth2nearprover does not exist creating it.
Created account eth2nearprover
Deployed contract to account eth2nearprover
Initializing client and prover contracts.
EthClient is not initialized, initializing...
connection not open on send()
(node:15984) UnhandledPromiseRejectionWarning: Error: connection not open
    at WebsocketProvider.send (/Users/maksymzavershynskyi/Projects/rainbow-bridge/environment/node_modules/web3/node_modules/web3-providers-ws/src/index.js:282:18)
    at RequestManager.send (/Users/maksymzavershynskyi/Projects/rainbow-bridge/environment/node_modules/web3/node_modules/web3-core-requestmanager/src/index.js:132:66)
    at sendRequest (/Users/maksymzavershynskyi/Projects/rainbow-bridge/environment/node_modules/web3/node_modules/web3-core-method/src/index.js:705:42)
    at Eth.send [as getBlockNumber] (/Users/maksymzavershynskyi/Projects/rainbow-bridge/environment/node_modules/web3/node_modules/web3-core-method/src/index.js:726:13)
    at Eth2NearClientContract.maybeInitialize (/Users/maksymzavershynskyi/Projects/rainbow-bridge/environment/lib/eth2near-client-contract/index.js:126:54)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:15984) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:15984) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
</details>

**Comment by ailisp** [comment]
This is also I think very important, and you list a very good plan!

**Comment by nearmax** [comment]
near-api-js with ECONRESET is getting more severe. eth-relayer is occasionally crashing, we cannot have a stable bridge until it is fixed.

**Comment by ailisp** [comment]
> near-api-js with ECONRESET is getting more severe. 

Is it been work around by catch and retrying in #156 ?

> eth-relayer is occasionally crashing, we cannot have a stable bridge until it is fixed.

In eth2near-relay All potential failing function calls is catched and return back to the top level `while(true)`, it should print error messages and continue retry. In near2eth-relay i didn't encounter a fail so didn't do a similar comprehensive error check, can be some missings that crash, will address them. (also let me know logs of any eth2near-relay, if there's any more crash cases to fix)

**Comment by nearmax** [comment]
Unfortunately, we initialize `ethClientContract` outside the loop and it holds persistent connection to the node AFAI which when it crashes prevents it from being reused.

**Comment by ailisp** [comment]
> Unfortunately, we initialize `ethClientContract` outside the loop and it holds persistent connection to the node AFAI which when it crashes prevents it from being reused.

```
this.ethClientContract = ethClientContract;
this.web3 = new Web3(ethNodeURL);
```
I see. web3 is keeping the persistent websocket conn, but I've observed after ECONRESET it can still have successful txn in devnet load test (need further confirmation, could also be i retried create a new connection). It does reuse connection here: https://github.com/near/near-api-js/blob/master/src/utils/web.ts#L21

**Comment by ailisp** [comment]
The four issues mentioned here are fixed in #156
